pkg/pillar/hypervisor/kubevirt.go

```diff
@@ -0,0 +1,1421 @@
+	// dispName is for vmi name/handle on kubernetes
+	meta := vmiMetaData{
+		vmi:      vmi,
+		name:     kubeName,
+		domainID: int(rand.Uint32()),
+	}
+	ctx.vmiList[domainName] = &meta
+
+	vmiStr := fmt.Sprintf("%+v", vmi)
+
+	// write to config file
+	file.WriteString(vmiStr)
+
+	return nil
+}
+
+func (ctx kubevirtContext) Start(domainName string) error {
+	logrus.Debugf("Starting Kubevirt domain %s", domainName)
+
+	err := getConfig(&ctx)
+	if err != nil {
+		return err
+	}
+	kubeconfig := ctx.kubeConfig
+
+	vmis, ok := ctx.vmiList[domainName]
+	if !ok {
+		return logError("start domain %s failed to get vmlist", domainName)
+	}
+	if vmis.isPod {
+		err := StartPodContainer(kubeconfig, ctx.vmiList[domainName].pod)
+		return err
+	}
+
+	vmi := vmis.vmi
+	virtClient, err := kubeapi.GetKubevirtClientSet(kubeconfig)
+
+	if err != nil {
+		logrus.Errorf("couldn't get the kubernetes client API config: %v", err)
+		return err
+	}
+
+	// Create the VM
+	i := 5
+	for {
+		_, err = virtClient.VirtualMachineInstance(kubeapi.EVEKubeNameSpace).Create(context.Background(), vmi)
+		if err != nil {
+			if strings.Contains(err.Error(), "dial tcp 127.0.0.1:6443") && i <= 0 {
+				logrus.Infof("Start VM failed %v\n", err)
+				return err
+			}
+			time.Sleep(10 * time.Second)
+			logrus.Infof("Start VM failed, retry (%d) err %v", i, err)
+		} else {
+			break
+		}
+		i = i - 1
+	}
+
+	logrus.Infof("Started Kubevirt domain %s", domainName)
+
+	err = waitForVMI(vmis.name, true)
+	if err != nil {
+		logrus.Errorf("couldn't start VMI %v", err)
+		return err
+	}
+
+	return nil
+
+}
+
+// Create is no-op for kubevirt, just return the domainID we already have.
+func (ctx kubevirtContext) Create(domainName string, cfgFilename string, config *types.DomainConfig) (int, error) {
+	return ctx.vmiList[domainName].domainID, nil
+}
+
+// There is no such thing as stop VMI, so delete it.
+func (ctx kubevirtContext) Stop(domainName string, force bool) error {
+	logrus.Debugf("Stop called for Domain: %s", domainName)
+	err := getConfig(&ctx)
+	if err != nil {
+		return err
+	}
+	kubeconfig := ctx.kubeConfig
+
+	vmis, ok := ctx.vmiList[domainName]
+	if !ok {
+		return logError("domain %s failed to get vmlist", domainName)
+	}
+	if vmis.isPod {
+		err := StopPodContainer(kubeconfig, vmis.name)
+		return err
+	} else {
+		virtClient, err := kubeapi.GetKubevirtClientSet(kubeconfig)
+		if err != nil {
+			logrus.Errorf("couldn't get the kubernetes client API config: %v", err)
+			return err
+		}
+
+		// Stop the VM
+		err = virtClient.VirtualMachineInstance(kubeapi.EVEKubeNameSpace).Delete(context.Background(), vmis.name, &metav1.DeleteOptions{})
+		if err != nil {
+			fmt.Printf("Stop error %v\n", err)
+			return err
+		}
+	}
+
+	if _, ok := ctx.vmiList[domainName]; ok {
+		delete(ctx.vmiList, domainName)
+	}
+
+	if _, ok := ctx.prevDomainMetric[domainName]; ok {
+		delete(ctx.prevDomainMetric, domainName)
+	}
+
```

Comment:
```suggestion
	delete(ctx.vmiList, domainName)

	delete(ctx.prevDomainMetric, domainName)

```

Reply:
Does delete check for existence ?

Reply:
The delete from a map only deletes if the key exists in the map.

Reply:
ok will fix that